### PR TITLE
New case study for the WebVR team

### PR DESCRIPTION
**Pull request by wibblymat**

### CODE REVIEW

**Comment by petele** [comment]
@jpmedley - would you please take a look.

File: src/content/en/showcase/2017/matterport2.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
+project_path: /web/_project.yaml
+book_path: /web/showcase/_book.yaml
+
+{# wf_published_on: 2017-06-12 #}
+{# wf_updated_on: 2017-06-12 #}
+{# wf_featured_image: /web/showcase/2017/images/matterport/top.jpg #}
+{# wf_featured_snippet: At Matterport, we have a large library of over 450,000 3D Spaces that users can find embedded in sites all over the web. Users could even experience these Spaces in VR through a native mobile app for Cardboard or Gear VR. Now with WebVR support, it is much easier for users to experience VR right from their browser.<br/><br/>Launching WebVR support in our player required us to focus on the most essential features. This case study describes the process we followed for optimizing for performance. We're very happy with the final result that we achieved and we're excited about all the things we can do in the future with WebVR. #}
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: 'We're very happy' is a little week. Maybe 'We're excited' or 'We're proud of'.

File: src/content/en/showcase/2017/matterport2.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+<img src="/web/showcase/2017/images/matterport/top.jpg" class="attempt-right">
+
+Matterport has <a href="https://matterport.com/gallery/">over 450,000 3D Spaces of buildings and
+locations</a> that users can find embedded in sites all over the web. The engineering team is
+always looking to make the experience of exploring our Spaces even more immersive, so VR is a high
+priority for us. When it became clear that WebVR was close to launching, we knew we had to mobilize
+as soon as possible to add support for it.
+
+## The Battle-Plan
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Headings 2, 3, etc. should be sentence case.

File: src/content/en/showcase/2017/matterport2.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+Our WebGL player has a lot of different bells and whistles: spatial annotations, automatic tours,
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: 'Our' in 'Our WebGL player' may be mistaken for Google. This needs to be changed throughout to 'The Matterport WebGL player'.

Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Suggestion: 'has a lot of different' change to 'supports varied'. (It's both shorter and stronger.)

File: src/content/en/showcase/2017/matterport2.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+multiple viewing modes and settings, and more. So when a team of three engineers had only four
+weeks before WebVR launched, we had to get creative. Instead of simply switching to VR mode in our
+current WebGL player, we decided to load a minimal, special WebVR player when the user initiates VR
+mode, and swap to that.
+
+This way we could cut the experience down to a minimal core experience of walking around inside a
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Need a comma between minimal and core.

File: src/content/en/showcase/2017/matterport2.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+## Switching to VR mode
+
+The easiest way we could think of for someone to enter VR was to detect when someone puts their
+phone in a VR headset while having our WebGL player open. This can be done by listening for the
+`vrdisplayactivate` event, which Chrome triggers when a user puts the phone in the Daydream headset
+and the NFC chip is detected.
+
+However, there is a small, but important difference between `vridisplayactivate` and other
+user-initiated events, e.g. onclick: Since we have to unload our normal WebGL player and load the
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: This has a bunch of problems, so here's what it needs to be:

user-initiated events (e.g. `onclick`). Since

File: src/content/en/showcase/2017/matterport2.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+WebVR player once we detect the user wanting to go into VR, we need a second or two of loading.
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: This whole sentence is clunky. Maybe:

When we detect a user going into VR, we need to unload the normal WebGL player and load the WebVR player. That takes a second or two.

File: src/content/en/showcase/2017/matterport2.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+Using `onclick` we are able to load this in the background during the controller calibration
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Change 'load this' to 'load WebVR'.

File: src/content/en/showcase/2017/matterport2.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+screen.  With `vrdisplayactivate`, though, we do not get to start loading until after the
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: 'during the controller calibration screen' is an odd phrase. Maybe 'during controller calibration'.

File: src/content/en/showcase/2017/matterport2.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+calibration, which ends up stalling the main thread when the user has the headset on, causing
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Change 'which ends up stalling' to 'which stalls'.

File: src/content/en/showcase/2017/matterport2.md [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+<a href="https://www.matterport.com/virtual-reality/webvr/"><img
+src="/web/showcase/2017/images/matterport/timeline.png" class="attemptfg-right"></a>
+
+Until `vrdisplayactivate` fires earlier, using `onclick` and a button provides users with the best
+experience for now.
+
+## Performance
+
+We knew performance would have to be rock solid for the experience to work well, so we started with
+the assumption that we would only implement a bare bones experience: no bells and whistles, just
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: 'so we started with the assumption that we would only implement a bare bones experience:' 

This is a bit wordy. Maybe:

'so we implemented a bare bones experience:'

File: src/content/en/showcase/2017/matterport2.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+simple navigation inside a space.
+
+We quickly made a prototype with <a
+href="https://github.com/borismus/webvr-boilerplate">webvr-boilerplate</a> and a stripped down
+version of our loading and rendering code. In our first run we only got about 10-20 FPS: In our
+normal WebGL app, that would feel bad, but in VR it’s downright nauseating.  As we often do, we
+fired up the Chrome Developer Tools, and connected it to a Pixel phone. We got four major
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Remove 'the' from Chrome Developer Tools.

File: src/content/en/showcase/2017/matterport2.md [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+takeaways:
+
+- Know your rendering code!
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Remove the exclamation point.

File: src/content/en/showcase/2017/matterport2.md [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    - Webvr-boilerplate rendered the screen a total of four times, two for each eye. Removing one
+      extraneous line of code quickly halved our rendering time! This has since been fixed in the
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Remove the word 'quickly' and the exclamation point.

File: src/content/en/showcase/2017/matterport2.md [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+      newest version of webvr-boilerplate, but we suspect that this particular bug has caused a lot
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Remove the word 'newest'.

File: src/content/en/showcase/2017/matterport2.md [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+      of developers to get the wrong impression of webvr performance.
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: 'webvr' to 'WebVR'

File: src/content/en/showcase/2017/matterport2.md [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+the assumption that we would only implement a bare bones experience: no bells and whistles, just
+simple navigation inside a space.
+
+We quickly made a prototype with <a
+href="https://github.com/borismus/webvr-boilerplate">webvr-boilerplate</a> and a stripped down
+version of our loading and rendering code. In our first run we only got about 10-20 FPS: In our
+normal WebGL app, that would feel bad, but in VR it’s downright nauseating.  As we often do, we
+fired up the Chrome Developer Tools, and connected it to a Pixel phone. We got four major
+takeaways:
+
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Bullets inside of bullets are rather ugly.  I'd do this:

**Know your rendering code!**

Webvr-boilerplate rendered the screen a total of four times, two for each eye. Removing one extraneous line of code quickly halved our rendering time! This has since been fixed in the newest version of webvr-boilerplate, but we suspect that this particular bug has caused a lot  of developers to get the wrong impression of webvr performance.

File: src/content/en/showcase/2017/matterport2.md [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+- Know your rendering code!
+    - Webvr-boilerplate rendered the screen a total of four times, two for each eye. Removing one
+      extraneous line of code quickly halved our rendering time! This has since been fixed in the
+      newest version of webvr-boilerplate, but we suspect that this particular bug has caused a lot
+      of developers to get the wrong impression of webvr performance.
+- Profile on the actual device you are building for, not your developer computer
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Maybe 'Profile on a typical device'. For the web, nobody should be targeting a single device.

File: src/content/en/showcase/2017/matterport2.md [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+    - Mobile devices and browsers have very different performance characteristics from your Mac;
+      make sure you’re optimizing the real bottleneck.
+- You don’t need all those pixels
+    - Be aware how many pixels you are rendering. In THREE.js, this is set with
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: 'Be aware' to 'Be aware of'

Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: What's THREE.js? Can you explain this without a reference to some random file in an unnamed framework? Alternatively you could start this 'For example, in the Wonderwidgets framework we edited THREE.js ...'

File: src/content/en/showcase/2017/matterport2.md [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+      WebGLRenderer.setPixelRatio(). You might not need the full resolution, in our case we simply
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Needs backticks.

Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: 'In our case' is the start of a new sentence.

File: src/content/en/showcase/2017/matterport2.md [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+      split our scene into smaller cubes and raycast into that, but you can also consider an
+      Octree.
+
+After a couple of rounds of improvements, we are now seeing super smooth experience on both the
+Pixel and the Pixel XL.
+
+## Living on the Bleeding Edge
+
+When you’re making something with a browser feature that is this new, you’re bound to see a lot of
+changes. In the WebVR specification, there was a mention of a security feature for iframes, which
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: remove the comma after 'iframes'.

File: src/content/en/showcase/2017/matterport2.md [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+embedders would have to add to their embed code to allow WebVR. This is important to us since our
+WebGL player is embedded in iframes all around the web. The new property looked something like
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Change 'The new property looked something like this:' to

'The new property, we thought, looked something like this:'

(You want to telegraph that what you're about to show is incorrect.)

File: src/content/en/showcase/2017/matterport2.md [hunk 22]
```diff
@@ -0,0 +1,122 @@
...
+this:
+
+    <iframe src="..." allowfullscreen allowvr>
+
+We started changing our documentation and embed codes that we hand out to customers, but wanted to
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Clunky sentence try:

'We started to change our documentation and code, but decided to ask the Chrome team if this was really needed.'

File: src/content/en/showcase/2017/matterport2.md [hunk 23]
```diff
@@ -0,0 +1,122 @@
...
+check with the Chrome team that this was really needed. **It wasn’t!** The Chrome team quickly
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: What I'm about to describe is a judgement call. Contrary to popular belief it is occasionally permissible to have a single sentence paragraph. It's done for emphasis. I would remove the bolding and exclamation point and make this it's own paragraph, then add the next sentence to the following paragraph.

File: src/content/en/showcase/2017/matterport2.md [hunk 24]
```diff
@@ -0,0 +1,122 @@
...
+entirely from the specification.
+
+## Conclusion
+
+<a href="https://www.matterport.com/virtual-reality/webvr/"><img
+src="/web/showcase/2017/images/matterport/end.jpg" class="attempt-right"></a>
+
+After a few expected bumps in the road, <a
+href="https://www.matterport.com/virtual-reality/webvr/">Matterport WebVR</a> was launched,
+delivering our VR Spaces to a larger audience directly from their browser, in a smooth, seamless
```
Comment (jpmedley) [inline src/content/en/showcase/2017/matterport2.md]: Remove the first comma.

**Comment by petele** [comment]
@wibblymat - when this is ready & you've made the updates, please remove the DO NOT MERGE flag and I'll take a look again, but will ignore in the mean time.

File: src/content/en/showcase/2017/matterport2.md (hunk 1 above)
Comment (wibblymat) [inline src/content/en/showcase/2017/matterport2.md]: @jpmedley it's a Matterport product term, and it is just 'Spaces'.

Comment (wibblymat) [inline src/content/en/showcase/2017/matterport2.md]: The whole article is written by Matterport, with their byline. Note the start of the summary:
> At Matterport, we have...

Does this whole summary need redoing in Google voice?

File: src/content/en/showcase/2017/matterport2.md (hunk 18 above)
Comment (wibblymat) [inline src/content/en/showcase/2017/matterport2.md]: THREE.js is a very popular and well known framework for doing 3D graphics on the web. THREE.js is to WebVR what jQuery or React are to PWAs. Have reworded slightly anyway.

**Comment by petele** [comment]
@wibblymat @jpmedley - what's the status on this?

**Comment by petele** [comment]
Hey folks - what's the status on this?

**Comment by petele** [comment]
@wibblymat - would you please update the status on this, or potentially close it?

**Comment by wibblymat** [comment]
Hey @petele, the status on this is that I have dropped the ball - I'll try and have this sorted ASAP

**Comment by ebidel** [comment]
Can we add `{# wf_blink_components: Blink>WebVR #}` to the metadata so it gets picked up on chromestatus?
